### ISSUE TITLE
isunit is wrong for polynomials over a non-integral domain

### ISSUE THREAD

**Issue by wbhart** [Nemocas/Nemo.jl#323]
Polynomials can be units in more cases than currently being tested, over a non-integral domain.

**Comment by wbhart** [comment]
We have the following theorems:

Thm: If R is of finite characteristic (and commutative) then f in R[x] is a unit iff f = u + r(x)*x for u a unit in R and all coefficients of r(x) are nilpotent in R.

Thm: a in Z/nZ is nilpotent iff rad(a) = rad(n).

**Comment by fingolfin** [comment]
In fact we don't need a check for the characteristic either. See e.g. https://kconrad.math.uconn.edu/blurbs/ringtheory/polynomial-properties.pdf

For `is_nilpotent`, we can implement generically for ...
- anything where `is_domain_type` holds
- univariate polynomial rings by delegating to the coefficient ring (a polynomial in $R[x]$ is nilpotent iff its coefficients are nilpotent in $R$
- this extends to multivariate rings (by considering $R[x,y]$ as $R[x][y]$), see Theorem 3.1 in the linked PDF
- ... and thus to `UniversalPolyRing`
- `ResElem{<:IntegerUnion}`
- we already handle `MatrixElem` in AA (though possibly with room for improvement?), so we have matrix rings covered

Dunno about Laurent and series rings (mainly because I am too lazy to think about them,  but my guess would be that at least for the Laurent rings it against reduces to considering the coefficients -- it shouldn't be too hard to either adapt the proofs or else follow them to construct a counter example)

**Comment by JohnAAbbott** [comment]
The article by Conrad excludes polynomials over the zero ring: presumably all such polynomials are both nilpotent and invertible.

**Comment by thofma** [comment]
Yes, all elements in the zero-ring are nilpotent and invertible. (I think both Theorem 2.1 and 2.2 are valid for zero rings; the `is_unit` method proposed by Max above also works independent of whether the coefficient ring is the zero ring.)

**Comment by JohnAAbbott** [comment]
Is there a general way of indexing/iterating over the terms of a "generalized polynomial": univariate, multivariate, polynomials, laurent polynomials, power-series, puiseux-series?
For the function `coeff`: I note that indexing over univariate polynomials uses indexes starting from 0 with no upper limit, whereas indexes into a multivariate polynomial start from 1 and must not exceed `length(f)`.
For a univariate puiseux-series one can use `S.scale` and `precision(S)` to find an upper limit...
Help/guidance would be appreciated!

**Comment by thofma** [comment]
No, there is no uniform interface. I would also avoid inexact rings for the moment.

**Comment by JohnAAbbott** [comment]
I wanted to implement `is_unit` and `is_nilpotent` for puiseux-series, but now realise that the definition is not entirely clear. Is the power series `z + O(z^2)` nilpotent?  If the ring is "capped" then it clearly is; if the ring is "relative" then no finite power will give zero.  Should the result depend on whether the ring is "capped"?

**Comment by thofma** [comment]
Yes, this is why we should not implement it. It is not clear what the semantics of equality is or should be. At the moment one gets a "no such method exists". Which is fine, since it is not a wrong result. On the other hand, for polynomials the functions are currently wrong.

**Comment by JohnAAbbott** [comment]
If the `base_ring` is the zero ring then everything is both nilpotent and invertible.
We can in some cases decide if a *-series is invertible or not; in other cases we do not have enough information.  So should the code attempt to decide if possible, and otherwise give a _not enough info_ error?  Similarly, in some cases it is clear that a *-series is not nilpotent, but we can never decide that it is nilpotent...

**Comment by fingolfin** [comment]
@JohnAAbbott I would suggest to ignore the series case completely for now; just let it throw a "not implemented" error for now. Instead focus on the cases we can handle. Once those are in, we have covered the most pressing need, and then can take our time to figure out the other times one at a time.

**Comment by thofma** [comment]
They should go (in AbstractAlgebra) in `src/Poly.jl` and `src/MPoly.jl` respectively.

**Comment by JohnAAbbott** [comment]
I also have implementations for Laurent polys.  It seems that `src/LaurentPoly.jl` contains very little, while there are further functions/operations in `src/algorithm/LaurentPoly.jl`; **in contrast** `src/LaurentMPoly.jl` contains several functions, and there is no file `src/algorithm/LaurentMPoly.jl`.  This is a slightly incoherent design structure.  Currently I'm adding the new functions to `src/algorithm/LaurentPoly.jl` and to `src/LaurentMPoly.jl`...
